casper/src/main/scala/coop/rchain/casper/MultiParentCasperImpl.scala

```diff
@@ -66,29 +62,26 @@ class MultiParentCasperImpl[F[_]: Sync: Capture: ConnectionsCell: TransportLayer
   private val faultToleranceThreshold         = 0f
   private val lastFinalizedBlockHashContainer = Ref.unsafe[F, BlockHash](genesis.blockHash)
 
-  private val processingBlock          = new SyncVar[Unit]()
-  private val PROCESSING_BLOCK_TIMEOUT = 5 * 60 * 1000L
-  processingBlock.put(())
-
   def addBlock(b: BlockMessage): F[BlockStatus] =
-    for {
-      _              <- Sync[F].delay(processingBlock.take(PROCESSING_BLOCK_TIMEOUT))
-      dag            <- blockDag
-      blockHash      = b.blockHash
-      containsResult <- dag.contains(blockHash)
-      result <- if (containsResult || blockBuffer.exists(
-                      _.blockHash == blockHash
-                    )) {
-                 Log[F]
-                   .info(
-                     s"Block ${PrettyPrinter.buildString(b.blockHash)} has already been processed by another thread."
-                   )
-                   .map(_ => BlockStatus.processing)
-               } else {
-                 internalAddBlock(b)
-               }
-      _ <- Sync[F].delay(processingBlock.put(()))
-    } yield result
+    Sync[F].bracket(semaphore.acquire)(
+      _ =>
+        for {
+          dag            <- blockDag
+          blockHash      = b.blockHash
+          containsResult <- dag.contains(blockHash)
+          result <- if (containsResult || blockBuffer.exists(
+                          _.blockHash == blockHash
+                        )) {
+                     Log[F]
+                       .info(
+                         s"Block ${PrettyPrinter.buildString(b.blockHash)} has already been processed by another thread."
+                       )
+                       .map(_ => BlockStatus.processing)
+                   } else {
+                     internalAddBlock(b)
+                   }
+        } yield result
+    )(_ => semaphore.release)
```

Comment:
This idiom `F.bracket(acquire)(_ => t)(release)` is called `Semaphore.withPermit`.